### DISCUSSION TITLE
💡 Require Dependency Injection in CSLA 6

### DISCUSSION THREAD

**Discussion by rockfordlhotka** [MarimerLLC/csla#1603] [General]
For CSLA 6 I am considering requiring that some dependency injection service be enabled/configured in all apps.

(from a [thread in the old forum](https://github.com/MarimerLLC/cslaforum/issues/948))

Microsoft has included (and used) dependency injection in ASP.NET Core and will continue in .NET 5.

Microsoft has _enabled_ the use of dependency injection for Windows Forms, WPF, console apps, etc. Their templates don't set it up, but it isn't hard to do.

The reason I'm considering requiring it for CSLA 6 is that if I _know_ that I can rely on there being a service provider it becomes much easier (and sometimes just plain _possible_) to implement some things throughout CSLA.

For example, it would be ideal if you could inject the current logging component into the data portal so you could log events. Today that's possible, but not as easy as in ASP.NET Core, because I don't know for sure that there's a service provider.

There are all sorts of other scenarios where CSLA _could_ do nice/helpful things if only we knew for sure that a service provider was available.

The only negative/breaking impact here, is that any Windows Forms, WPF, console, or other not-ASP.NET style apps would need to add a few lines of code into `Program.cs` or the startup equivalent file. So I don't think the burden is high, and the potential gains are _very high_.

Thoughts?

**Comment by Chicagoan2016** [comment]
I keep up with Csla's code using Samples (Project Tracker) for the most part, I am lazy :).
Since we couldn't do constructor injection with Csla business types, do we plan to use DI with Csla business types?
Kind Regards

**Reply by rockfordlhotka** [reply]
Right now in CSLA 5 you can do DI in business types where it matters: the data portal operation methods.

```c#
  [Fetch]
  private void Fetch(int id, [Inject] MyDataLayer dal)
  {
    var data = dal.GetData(id);
  }
```

There's no value in constructor injection for mobile types - where the instance might materialize and dematerialize on different physical tiers. At least no one has come up with a scenario where that makes sense yet.

**Comment by GillesBer** [comment]
This sounds great on any .NET Core application.

As CLSA6 will still support .NET 4.8, will this apply also to .NET 4.8 apps?
I know how to apply it on our 'modernized to .NET Core' app ex:
https://thecodebuzz.com/dependency-injection-windows-form-desktop-app-net-core/
but would that apply to our legacy 4.8 apps too?

**Reply by rockfordlhotka** [reply]
Yes, this will apply to .NET Framework apps as well. The DI framework is available to .NET Framework 4.7.2 and higher, so it should work fine for legacy apps.

You can ignore DI other than at app startup, where you'll have to initialize the `ServiceProvider`.
